Vision_feats/generate_feats.py

```diff
@@ -0,0 +1,308 @@
+        """Fetches the item at a given index.
+        """
+        img = None
+        path = self.dataset[index]
+        with open(path, 'rb') as f:
+            img = Image.open(f)
+            if img.mode != 'RGB':
+                self.num_grayscale += 1
+                img = img.convert('RGB')
+            img = self.transform(img)
+        return img
+
+    def __len__(self) -> int:
+        """Returns the number of items in the dataset.
+        """
+        l = len(self.dataset)
+        return l
+
+
+class MyResNet50(nn.Module):
+    def __init__(self, args):
+        """Initialize network layers.
+        """
+        super().__init__()
+        model = resnet50(pretrained=True)
+        self.conv_layers = nn.Sequential(*list(model.children())[:-1])
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Perform the forward pass with the net. Just output the feature.
+        """
+        model_output = self.conv_layers(x)
+        return model_output
+
+
+class MyBeit(nn.Module):
+    def __init__(self, args):
+        """Initialize network layers.
+        """
+        super().__init__()
+        #self.feat_ex = BeitFeatureExtractor.from_pretrained("microsoft/beit-large-patch16-224")
+        self.model = BeitModel.from_pretrained(
+            models_file[args.img_encode_model]
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Perform the forward pass with the net. Just output the feature.
+        """
+        #x = self.feat_ex(x)
+        x = {'pixel_values': x}
+        model_output = self.model(**x)
+        model_output = model_output.last_hidden_state
+        return model_output
+
+
+class MyClip(nn.Module):
+    def __init__(self, args):
+        """Initialize network layers.
+        """
+        super().__init__()
+        #self.feat_ex = BeitFeatureExtractor.from_pretrained("microsoft/beit-large-patch16-224")
+        full_model, _ = clip.load(models_file[args.img_encode_model]) 
+        #self.model = full_model.visual
+        self.model = full_model
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Perform the forward pass with the net. Just output the feature.
+        """
+        #x = self.feat_ex(x)
+        model_output = self.model.encode_image(x).float()
+        return model_output
+
+
+class VisionEncode(nn.Module):
+    def __init__(self, args):
+        """A vision encoder for processing images on the fly
+        """
+        super().__init__()
+        # Warning: The feature is resized and cropped before
+
+        self.models_dict = {
+            'resnet': MyResNet50,
+            'beit': MyBeit,
+            'beit_ft': MyBeit,
+            'clip': MyClip,
+            'clipL': MyClip
+        }
+        # Load model
+        self.model = self.models_dict[args.img_encode_model](args)
+        self.args = args
+        if self.args.downsample == 'transformer':
+            self.adaptor = nn.Transformer(nhead=8, num_encoder_layers=6)
```

Comment:
Less hard-coding of arguments here and below :)

Reply:
Yeah, the hard-coding thing is just for quick experiments :) So the caveat here is we actually don't use any of these downsampling methods except BEIT... and I am thinking of putting BEIT aside for now since it's not really important here. I probably would spend more time on this later if we need it. For now, does commenting out the whole downsample thing sound good to you?